### ISSUE TITLE
openseq2seq big transformer comparison to fairseq big transformer

### ISSUE THREAD

**Issue by vsuthichai** [NVIDIA/OpenSeq2Seq#250]
I'm currently benchmarking the big transformer of openseq2seq and fairseq, and I'm focusing on getting the single gpu throughput performance of Openseq2seq to that of fairseq's.  I eventually will go to multiple nodes, but I wanted to work through this with a simpler setting.

OpenSeq2Seq big transformer single gpu mixed precision training with 128bsz, 16gb volta, I get about 350ms per step.
Fairseq big transformer single gpu fp16 training with 3584 max tokens per batch, 16gb volta, I get about 148ms per step.

So, fairseq's implementation is still better by over 2x.  Their forward and backward pass is faster.  I'm also attaching output of profiling with nvprof.  This was generated over 40 steps of training on both Fairseq and OpenSeq2Seq.  Looking at the top of each profile which is sorted in descending order by execution time, fairseq implementation seems to make more efficient use of fp16 s884 gemm calls.

[fairseq.txt](https://github.com/NVIDIA/OpenSeq2Seq/files/2474838/fairseq.txt)
[openseq2seq.txt](https://github.com/NVIDIA/OpenSeq2Seq/files/2474840/openseq2seq.txt)

**Comment by vsuthichai** [comment]
Typos, sorry

**Comment by cbockman** [comment]
@vsuthichai did you resolve this?  What was the outcome?

**Comment by vsuthichai** [comment]
@cbockman I never did fully resolve this.

At the time I opened the ticket, I was using 128 sentence pair batch size and getting about 350ms per step on a single gpu V100 with OpenSeq2Seq's big transformer.

SInce then,  I've switched out the dataset pipeline from ParallelTextLayer to the TransformerDataLayer and configured the batch size (now measured in tokens) to be 3584, the same as fairseq's, simplified certain parts of the network like the removal and re-adding of padding, ensured axis sizes were properly padded to multiples of 8, grouped sentences by length, etc.  This brought the time down to an average of 218ms per step.  It is still however short of fairseq's ~150ms.

Looking at the nvprof output, it seems that fairseq/pytorch has implemented their ops more efficiently than tensorflow, at least in the context of the big transformer.

With that being said, there are some small differences in the network architecture however.  Fairseq does the layer norm in fp16 which provides a tiny amount of speedup, while OpenSeq2Seq performs a safe layer norm calculation in fp32.  Fairseq also adheres more strictly to the Vaswani paper.  The two networks are not equivalent.
